exporters/prometheus/src/exporter_utils.cc

```diff
@@ -167,6 +167,156 @@ std::string PrometheusExporterUtils::SanitizeNames(std::string name)
+                                                  {"Cel", "celsius"},
+                                                  {"Hz", "hertz"},
+                                                  {"1", ""},
+                                                  {"%", "percent"},
+                                                  {"$", "dollars"}};
+  auto res_it = units.find(unit_abbreviation);
+  if (res_it == units.end())
+  {
+    return unit_abbreviation;
+  }
+  return res_it->second;
+}
+
+std::string PrometheusExporterUtils::GetPrometheusPerUnit(const std::string &per_unit_abbreviation)
+{
+  static std::map<std::string, std::string> per_units{
+      {"s", "second"}, {"m", "minute"}, {"h", "hour"}, {"d", "day"},
+      {"w", "week"},   {"mo", "month"}, {"y", "year"}};
+  auto res_it = per_units.find(per_unit_abbreviation);
+  if (res_it == per_units.end())
+  {
+    return per_unit_abbreviation;
+  }
+  return res_it->second;
+}
+
+std::string PrometheusExporterUtils::RemoveUnitPortionInBraces(const std::string &unit)
+{
+  return std::regex_replace(unit, CHARACTERS_BETWEEN_BRACES_PATTERN, "");
+}
+
+std::string PrometheusExporterUtils::ConvertRateExpressedToPrometheusUnit(
+    const std::string &rate_expressed_unit)
+{
+  if (rate_expressed_unit.find("/") == std::string::npos)
+  {
+    return rate_expressed_unit;
+  }
+
+  std::vector<std::string> rate_entities;
+  size_t pos = rate_expressed_unit.find("/");
+  rate_entities.push_back(rate_expressed_unit.substr(0, pos));
+  rate_entities.push_back(rate_expressed_unit.substr(pos + 1));
+
+  if (rate_entities[1].empty())
+  {
+    return rate_expressed_unit;
+  }
+
+  std::string prometheus_unit     = GetPrometheusUnit(rate_entities[0]);
+  std::string prometheus_per_unit = GetPrometheusPerUnit(rate_entities[1]);
+
+  return prometheus_unit + "_per_" + prometheus_per_unit;
+}
+
+std::string PrometheusExporterUtils::CleanUpString(const std::string &str)
+{
+  std::string cleaned_string = std::regex_replace(str, INVALID_CHARACTERS_PATTERN, "_");
+  cleaned_string = std::regex_replace(cleaned_string, SANITIZE_CONSECUTIVE_UNDERSCORES, "_");
+  cleaned_string = std::regex_replace(cleaned_string, SANITIZE_TRAILING_UNDERSCORES, "");
+  cleaned_string = std::regex_replace(cleaned_string, SANITIZE_LEADING_UNDERSCORES, "");
+
+  return cleaned_string;
+}
+
+std::string PrometheusExporterUtils::MapToPrometheusName(
+    const std::string &name,
+    const std::string &unit,
+    prometheus_client::MetricType prometheus_type)
+{
+  auto sanitized_name                    = SanitizeNames(name);
+  std::string prometheus_equivalent_unit = GetEquivalentPrometheusUnit(unit);
+
+  // Append prometheus unit if not null or empty.
+  if (!prometheus_equivalent_unit.empty() &&
+      sanitized_name.find(prometheus_equivalent_unit) == std::string::npos)
+  {
+    sanitized_name += "_" + prometheus_equivalent_unit;
+  }
+
+  // Special case - counter
+  if (prometheus_type == prometheus_client::MetricType::Counter &&
```

Comment:
Counters MUST end in _total, so we need to be less tolerant here.  We can only skip the _total suffix if the name **ends** in _total.  Since we need unit suffixes to come before type suffixes, one approach would be to trim _total suffixes from the name before appending the unit suffix, and finally append the final _total suffix.

Reply:
thanks, will not add total unless if it has already.